proto-lens-protoc/src/Data/ProtoLens/Compiler/Definitions.hs

```diff
@@ -78,32 +79,45 @@ data Definition n = Message (MessageInfo n) | Enum (EnumInfo n)
 data MessageInfo n = MessageInfo
     { messageName :: n  -- ^ Haskell type name
     , messageDescriptor :: DescriptorProto
-    , messageFields :: [FieldInfo]
-      -- ^ The Haskell names for each field.
+    , messageFields :: [FieldInfo] -- ^ Fields not belonging to a oneof.
     , messageOneofFields :: [OneofInfo]
-      -- This list corresponds 1-1 with "field" in messageDescriptor.
+      -- ^ The oneofs in this message, associated with the fields that
+      --   belong to them.
     } deriving Functor
 
 -- | Information about a single field of a proto message.
 data FieldInfo = FieldInfo
-    { overloadedField :: String
-      -- ^ The Haskell overloaded name of this field; may be shared between two
-      -- different message data types.
-    , recordFieldName :: Name
-      -- ^ The Haskell name of this internal record field.  Unique within each
-      -- module.
-    , fieldDescriptor :: FieldDescriptorProto
-    , oneofFieldInfo :: Maybe OneofFieldInfo
+    { fieldDescriptor  :: FieldDescriptorProto
+    , plainFieldName :: FieldName
     }
 
 data OneofInfo = OneofInfo
-    { oneofTypeName :: String
-    , oneofRecordFieldName :: Name
+    { oneofFieldName :: FieldName
+    , oneofTypeName :: Name
+      -- ^ The name of the sum type corresponding to this oneof.
+    , oneofCases :: [OneofCase]
+      -- ^ The individual fields that make up this oneof.
+    }
+
+data OneofCase = OneofCase
+    { caseField :: FieldInfo
+    , caseConstructorName :: Name
+        -- ^ The constructor for building a 'oneofTypeName' from the
+        -- value in this field.
     }
 
-data OneofFieldInfo = OneofFieldInfo
-    { oneofConstructorName :: Name
-    , oneofEnclosingFieldName :: Name
+data FieldName = FieldName
+    { overloadedName :: String
```

Comment:
At some points (e.g., when we promote it to a type-level string) we actually need a raw `String`.  (Also, for some lenses we want to prefix it with `"maybe'"`.  The `Name` datatype is awkward to use for those cases.

Added a "Symbol" newtype to make the meaning more explicit.  (It should also help prevent some bugs that I ran into when developing this PR.)